Bugzilla/JobQueue/Runner.pm

```diff
@@ -173,40 +178,63 @@ sub gd_check {
     print get_text('job_queue_depth', { count => $count }) . "\n";
 }
 
+# override this to use IO::Async.
 sub gd_setup_signals {
-    my $self = shift;
-    $self->SUPER::gd_setup_signals();
-    $SIG{TERM} = sub { $self->gd_quit_event(); }
+    my $self    = shift;
+    my %signals = (
+        # Daemon::Generic by default handles these,
+        INT  => sub { $self->gd_quit_event() },
+        HUP  => sub { $self->gd_reconfig_event() },
+        # Bugzilla adds this
+        TERM => sub { $self->gd_quit_event() },
+    );
+    my $loop = IO::Async::Loop->new;
+    foreach my $name (keys %signals) {
+        my $signal = IO::Async::Signal->new(
+            name       => $name,
+            on_receipt => $signals{$name},
+        );
+        $loop->add($signal);
+    }
 }
 
 sub gd_other_cmd {
     my ($self) = shift;
     if ($ARGV[0] eq "once") {
-        $self->_do_work("work_once");
-
-        exit(0);
+        exit $self->run_worker("work_once")->get;
     }
-    
+
     $self->SUPER::gd_other_cmd();
 }
 
 sub gd_run {
     my $self = shift;
 
-    $self->_do_work("work");
+    exit $self->run_worker("work")->get;
```

Comment:
you should `unlink($self->{gd_pidfile})` before exiting to match `Daemon::Generic`'s current behaviour.